static/app/components/draggableTabs/draggableTabList.tsx

```diff
@@ -24,6 +38,202 @@ interface BaseDraggableTabListProps extends DraggableTabListProps {
   items: DraggableTabListItemProps[];
 }
 
+function useIsDragging(ref: React.RefObject<HTMLElement>) {
+  const [isDragging, setIsDragging] = useState(false);
+
+  useEffect(() => {
+    const element = ref.current;
+
+    if (!element) {
+      return () => {};
+    }
+
+    const handlePointerDown = () => {
+      setIsDragging(true);
+    };
+    const handlePointerUp = () => {
+      setIsDragging(false);
+    };
+    element.addEventListener('pointerdown', handlePointerDown);
+    document.addEventListener('pointerup', handlePointerUp);
+
+    return () => {
+      element.removeEventListener('pointerdown', handlePointerDown);
+      document.removeEventListener('pointerup', handlePointerUp);
+    };
+  }, [ref]);
+
+  return isDragging;
+}
+
+function useOverflowingTabs({state}: {state: TabListState<DraggableTabListItemProps>}) {
+  const persistentTabs = [...state.collection].filter(
+    item => item.key !== 'temporary-tab'
+  );
+  const outerRef = useRef<HTMLDivElement>(null);
+  const addViewTempTabRef = useRef<HTMLDivElement>(null);
+  const [tabElements, setTabElements] = useState<Array<HTMLDivElement | null>>([]);
+  const {width: outerWidth} = useDimensions({elementRef: outerRef});
+  const {width: addViewTempTabWidth} = useDimensions({elementRef: addViewTempTabRef});
+  const tabsDimensions = useDimensionsMultiple({elements: tabElements});
+
+  const overflowingTabs = useMemo(() => {
+    const availableWidth = outerWidth - addViewTempTabWidth;
+
+    let totalWidth = 0;
+    const overflowing: Node<DraggableTabListItemProps>[] = [];
+
+    for (let i = 0; i < tabsDimensions.length; i++) {
+      totalWidth += tabsDimensions[i].width + 1; // 1 extra pixel for the divider
+      if (totalWidth > availableWidth + 1) {
+        overflowing.push(persistentTabs[i]);
+      }
+    }
+
+    return overflowing.filter(defined);
+  }, [outerWidth, addViewTempTabWidth, persistentTabs, tabsDimensions]);
+
+  return {
+    overflowingTabs,
+    setTabElements,
+    outerRef,
+    addViewTempTabRef,
+    persistentTabs,
+  };
+}
+
+function OverflowMenu({
+  state,
+  overflowTabs,
+}: {
+  overflowTabs: Node<DraggableTabListItemProps>[];
+  state: TabListState<any>;
+}) {
+  const options = useMemo(() => {
+    return overflowTabs.map(tab => {
+      return {
+        value: tab.key,
+        label: tab.textValue,
+        textValue: tab.textValue,
```

Comment:
Should this also include the tab menu button for selected, overflowed tabs? Right now there's no way to access that functionality for overflowing tabs other than by increasing screen width.

Side note: perhaps we should talk to Vu about what this should look like? I don't think there are any mockups for this

Reply:
I don't have it implemented in this PR, but I thought the intended behavior was that selecting a tab in the overflow menu would pop it out into the tab bar. That way we won't need to add those controls to the overflow. But yeah we never got any designs on the overflow menu specifically

Reply:
Gotcha, my concern with that would be that the tab ordering might get a little confusing. Would the tab replace the tab in the right-most tab that's not overflowing? I'll check in with Vu tomorrow to see what his thoughts are, but also agreed that this can be part of a different PR.